docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,813 @@
+Runtime Extension, you are required to implement an HTTP server handling requests according to the
+OpenAPI specification for the Runtime Hook you are interested in.
+
+Please note that Cluster API only enforces input and output parameters types as defined by a Runtime Hook version,
+and developers are fully responsible of all the other elements of the design of a Runtime Extension implementation,
+including:
+
+- To choose which programming language to use; please note that for sake of this proposal Golang is the language
+  of choice, and we are not planning to test/provide tooling/libraries for other languages. Nevertheless, given that
+  we rely on Open API and plain HTTP(s) calls, other languages should just work but support will be provided at
+  best effort.
+- In case it is required to implement more than one extension, to choose if to have a single HTTP server with multiple
+  handlers or separated HTTP servers;
+- To choose if to have a dedicated HTTP server(s) for Runtime Extensions only or if to use the HTTP server for other
+  purposes as well (e.g. metric endpoint).
+
+In case the Runtime Extension is being developed in Golang, the implementer can benefit from importing
+`sigs.k8s.io/cluster-api` and:
+
+- Use Golang types defined under `/runtime` (the types from which the OpenAPI specification has been generated).
+- Use the RuntimeExtension catalog object to generate the skeleton of the HTTP handler for a given Runtime extension.
+  The generated func will take care of scaffolding tasks like Marshal/Unmarshal; the only missing part will be to
+  implement a strongly typed func that contains the actual extension implementation.
+
+E.g.
+
+```go
+// Note: this is pseudo code, meant to demonstrate that implementing a RuntimeExtension requires only minimal scaffolding;
+// the exact details will be defined during implementation, possibly taking advantage of Golang generics.
+
+var c = catalog.NewCatalog()
+
+func init() {
+    v1alpha2.AddToCatalog(c)
+}
+
+func main() {
+    ctx := context.Background()
+
+    listener, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", "8082"))
+    if err != nil {
+        panic(err)
+    }
+
+    // Build an HTTP handler for a given operation, calling a strongly typed func at each request
+    beforeUgradeHandler, err := catalogHTTP.NewHandlerBuilder().
+        WithCatalog(c).
+        AddService(&v1alpha2.BeforeUgradeHook{}, doBeforeUpgrade).
+        Build()
+    if err != nil {
+        panic(err)
+    }
+
+    srv := &http.Server{
+        Handler: BeforeUgradeHandler,
+    }
+    if err := srv.Serve(listener); err != nil && err != http.ErrServerClosed {
+        panic(err)
+    }
+}
+
+func doBeforeUpgrade(in *v1alpha2.BeforeUpgradeInput, out *v1alpha2.BeforeUpgradeOutput) error {
+    // your actual implementation...
+    return nil
+}
+```
+
+While writing the actual code of the Runtime Extension a set of important guidelines must be considered:
+
```

Comment:
should we have an "errors" section? what happens when one or more extensions fail? how do we react to the different http response? Do we need a .failurePolicy in the API?

Reply:
I just saw .FailurePolicy is include already, still probably worth it an error section here.

Reply:
Is it ok if I iterate on this while we figure out how to handle HTTP return codes (we are investigating how this is implemented in the API server for dynamic admission webhooks)

Reply:
Yup would be good if we can iterate a bit on it. Overall we're trying to make everything more concrete over time when we figure out how to actually use the Runtime SDK to implement external patches / lifecycle hooks.

Reply:
IMO what happens when things fail is a critical part of the design and we shouldn't leave it for later. Seems like this proposal already is introducing opinions on the topic (failurePolicy, timeout) so it would make sense to summarize in an error section as @enxebre pointed out. We can list the stuff that is still TBD / future work in that section as well.

Reply:
@CecileRobertMichon @enxebre 
I have documented error management as well as how blocking runtime extensions should work